feat(common): add ability to retrieve the state from Location service

jasonaden opened the pull request:
Previously there wasn't a way to retrieve `history.state` from the `Location` service. The only time the framework exposed this value was in navigation events. This meant if you weren't using the Angular router, there wasn't a way to get access to this `history.state` value other than going directly to the DOM.

This PR adds an API to retrieve the value of `history.state`. This will be useful and needed to provide a backwards-compatible `Location` service that can emulate AngularJS's `$location` service since we will need to be able to read the state data in order to produce AngularJS location transition events.

This feature will additionally be useful to any application that wants to access state data through Angular rather than going directly to the DOM APIs.

jasonaden committed:
feat(common): add ability to retrieve the state from Location service

Previously there wasn't a way to retrieve `history.state` from the `Location` service. The only time the framework exposed this value was in navigation events. This meant if you weren't using the Angular router, there wasn't a way to get access to this `history.state` value other than going directly to the DOM.

This PR adds an API to retrieve the value of `history.state`. This will be useful and needed to provide a backwards-compatible `Location` service that can emulate AngularJS's `$location` service since we will need to be able to read the state data in order to produce AngularJS location transition events.

This feature will additionally be useful to any application that wants to access state data through Angular rather than going directly to the DOM APIs.

jasonaden commented:
[Presubmit](http://test/OCL:232751665:BASE:232751824:1549493532086:8fa33a8e)

alfaproject commented on packages/common/src/location/location.ts:
```diff
@@ -78,6 +78,11 @@ export class Location {
+  /**
+   * Returns the current value of the history.state object.
+   */
+  state(): any { return this._platformStrategy.getState(); }
```
If it's a method shouldn't it have a verb? Like `getState`

jasonaden replied:
I would tend to agree. I didn't do this because we already have `path()` method (a getter) without that naming convention. But I prefer to have `get` on there, so will go ahead and change it.

sod commented on packages/common/src/location/path_location_strategy.ts:
```diff
@@ -94,4 +94,6 @@ export class PathLocationStrategy extends LocationStrategy {
 
   back(): void { this._platformLocation.back(); }
+
+  getState(): any { return this._platformLocation.getState(); }
```
Being on typescript 3+, wouldn't it be way nicer for new apis to use `unkown` (see http://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html) instead of `any`? Or even better, `getState<T = unkown>(): T`?

The beauty of unkown is that typescript forces you to check the value at runtime as it may indeed be anything. `any` on the other hand disables all typescript intellisense alltogether.

jasonaden replied:
Yes! Thanks for the suggestion.

jasonaden replied:
Leaving out the generic on this function. Reasoning is it can allow implicit type inference:

```const foo: string = location.getState();```

Is basically an implicit cast. It defeats the point of using `unknown` as it will implicitly infer a value for `T` when used in a typed context.

jasonaden committed:
feat(common): add MockPlatformLocation to enable more robust testing of Location services

Prior to this change we had a MockLocationStrategy to replace the Path and Hash Location Strategies. However, there wasn't a good way to test the PlatformLocation which is needed for doing things such as setting history.state, using back()/forward(), etc.

jasonaden commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:232751665:BASE:237485958:1552072978391:f9499a35)

jasonaden commented:
[Updated presubmit](http://test/OCL:232751665:BASE:237505530:1552079434509:c68abb2d)

petebacondarwin commented:
This PR is very stale, and @jasonaden is no longer working on Angular.
I feel like this PR was somewhat speculative to fix some AngularJS integration scenarios. But since it has not been touched in over a year, it seems unlikely that the original motivation is still relevant.
